### DISCUSSION TITLE
ThreadX Porting

### DISCUSSION THREAD

**Discussion by xiongyu0523** [earlephilhower/arduino-pico#292] [Q&A]
Hi, I'm trying to get threadx support to arduino-pico. I need three symbol to support RTOS port. 

1. Reset vector table after main (ram_vector_table)
2. Systick isr handler (SysTick_Handler)
3. PendSv isr handler (PendSV_Handler)

I use the value above but it seems the system is not working. Not sure if i'm using the right value.

**Comment by xiongyu0523** [comment]
for other platform, arduino core will use systick and provide a callback symbol but i couldn't find in arduino-pico.

**Reply by earlephilhower** [reply]
Sorry, none of those symbols exist or are needed in the arduino-pico core.  I'm also not familiar with the ThreadX libs so not sure what exactly it's looking for.

You have full Pico SDK access, though, so if you know what they're doing and what they're supposed to hook into then you can easily write them yourself.  The reset vector is already there under a different name, of course, or else the chip wouldn't run properly.  Systick IRQ, no idea but I suppose it'd be driven by one of the alarms. 

Good luck!

**Reply by xiongyu0523** [reply]
Thanks for your quick answer. Normally OS will need override the weak symbol of Systick and PendSV ISR to support threading. I did a [port](https://github.com/xiongyu0523/Azure-RTOS-on-Raspberry-Pi-Pico-RP2040) of ThreadX to pico-SDK early this year (before 1.2.0), at that time, the reset vector is ram_vector_table, and isr_systick/isr_pendsv for systick and pendsv.  

Can you let me know what is the renamed reset vector in arduino-pico? (when you say `The reset vector is already there under a different name`). Thanks.

**Reply by earlephilhower** [reply]
That's very odd.  I don't rename anything in the PICO SDK and, in fact, use their linker file pretty much unmodified.

Are you sure those symbols exist in the plain SDK?  And, if so, are you including the proper header file where they're exposed as `extern "C" ...` so that the library code can see them?  It might just be a case of the symbol being there, but C not being made aware of it because no prototype was defined.

**Reply by earlephilhower** [reply]
I looked at your port, and the port itself seems to contain one of the symbols you're looking for:

https://github.com/xiongyu0523/Azure-RTOS-on-Raspberry-Pi-Pico-RP2040/blob/b019a1f3f3e7df5b50e9c0fee94bc91f1e02c046/tx_initialize_low_level.S#L168-L170

Also, `ram_vector_table` is still present in the link section.  Here's a snippet of a `map` I looked at at random:

````
.ram_vector_table
                0x0000000020000000       0xc0
 *(.ram_vector_table)
 .ram_vector_table
                0x0000000020000000       0xc0 /home/earle/Arduino/hardware/pico/rp2040/lib/libpico.a(runtime.c.obj)
                0x0000000020000000                ram_vector_table

.data           0x00000000200000c0     0x19fc load address 0x000000001000e408
````

I also don't see any `PendSV_Handler` or `SysTick_Handler` when I do a `grep` in the sdk directory, except for references to a specific chip and FreeRTOS (which is not used here)...

Did you depend on FreeRTOS in the base SDK?

**Reply by xiongyu0523** [reply]
When you build pico-sdk, PICO_NO_RAM_VECTOR_TABLE is set to 0, and PICO_NO_FLASH is also 0, right?
So in runtime.c (before main), below code will set VTOR to ram_vector_table which is 0x20000000 and enable user to use API to install hanlder for interrupt (such as USB). Yesterday when I try to read back VTOR in a blink example on rp2040, I get value of 0. I don't think what's wrong?

I have tried to pass either 0, ram_vector_table, or __vectors (0x10000100) to my low_level.s. None of them works.. May be the problem is on the PendSV or Systick side?

```
#if !(PICO_NO_RAM_VECTOR_TABLE || PICO_NO_FLASH)
    __builtin_memcpy(ram_vector_table, (uint32_t *) scb_hw->vtor, sizeof(ram_vector_table));
    scb_hw->vtor = (uintptr_t) ram_vector_table;
#endif
```

All I want to do is override the default `weak` symbol of PendSV and Systick execption handler. they used to be isr_pendsv and isr_systick in pico-sdk before 1.2.0. And after May there is a new file reanme_expections.h was added to rename them to CMSIS standard symobls. I did tried both isr_ or standard symbol. None of them seems works. 

```
#define isr_pendsv PendSV_Handler
#define isr_systick SysTick_Handler
```

**Reply by earlephilhower** [reply]
Check in https://github.com/earlephilhower/arduino-pico/tree/master/tools/libpico for the `libpico.a` build process.  I don't define either of those two macros, and the `ram_vector_table` symbol *does* exist in the final link.  It may not be visible in your C code, of course, w/o the `extern` comment.  Make sure you're using `extern "C"` if the lib is in a `.cpp` file, of course.

In general, does your ported lib build/run on native SDK v1.2.0?  Maybe the SDK itself has tweaked something in the 1.2 release?

**Reply by xiongyu0523** [reply]
Great questions. I have updated my previous project and verified on 1.2.0 pico-sdk today. Everything is fine.

1. To use the new cmsis compatbile macro in rename_exceptions.h, I need add cmsis_core to my project's CMake file to link this interface library. Can we add this in arduino-pico, i asusme there is no side effect. 

https://github.com/xiongyu0523/Azure-RTOS-on-Raspberry-Pi-Pico-RP2040/blob/5394fd8856d02423ea55a2c1db4e6a53752dc75e/CMakeLists.txt#L22

2. The vector table is still `ram_vector_table`. I'm confused because in your first post you said you can find this symbol in image but not in the second one.. while I'm doing test, i can see `ram_vector_table` is 0x20000000.. 

https://github.com/xiongyu0523/Azure-RTOS-on-Raspberry-Pi-Pico-RP2040/blob/5394fd8856d02423ea55a2c1db4e6a53752dc75e/tx_initialize_low_level.S#L124

**Reply by Andy2No** [reply]
> isr_systick

I don't claim to understand the topic, but it seems there are two places in the sdk that need to be looked at:

(Search results in pico sdk) https://github.com/raspberrypi/pico-sdk/search?q=isr_systick

In crt0.S, kilograham has added a comment about the changes in May:

https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_standard_link/crt0.S

> kilograham Add basic CMSIS core headers (#384)
>Latest commit b05d863 on 11 May
>Adds a new cmsis_core INTERFACE library, which provides the basic CMSIS macros/definitions, and RP2040.h and system_RP2040.h
By default the isr_ methods in the vector table are renamed to their CMSIS versions, but this can be overridden by defining PICO_CMSIS_RENAME_EXCEPTIONS=0

That file defines the vector table, using the definitions of those two symbols, so you'd either have to define them first, I guess, or override the contents of the table afterwards:

```
global __vectors, __VECTOR_TABLE
__VECTOR_TABLE:
__vectors:
.word __StackTop
.word _reset_handler
.word isr_nmi
.word isr_hardfault
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_svcall
.word isr_invalid // Reserved, should never fire
.word isr_invalid // Reserved, should never fire
.word isr_pendsv
.word isr_systick
```

In rename_exceptions.h, the symbols are defined as you said: https://github.com/raspberrypi/pico-sdk/blob/bfcbefafc5d2a210551a4d9d80b4303d4ae0adf7/src/rp2_common/cmsis/include/cmsis/rename_exceptions.h

```
#if PICO_CMSIS_RENAME_EXCEPTIONS
#define isr_nmi NMI_Handler
#define isr_hardfault HardFault_Handler
#define isr_svcall SVC_Handler
#define isr_pendsv PendSV_Handler
#define isr_systick SysTick_Handler
```

My only previous encounter with SysTick_Hander and PendSV_Handler was in  cnlohr /
minimal_stm32f303 

Stubs for those functions were defined as part of the project code, in stm32f30x_it.c :+1: 
https://github.com/cnlohr/minimal_stm32f303/blob/master/stm32f30x_it.c

It seems I had to add a line of code to get that to compile, or to compile without warnings (I forget which), referencing the function that SysTick_Handler calls:

```
void PendSV_Handler(void)
{
}

/**
  * @brief  This function handles SysTick Handler.
  * @param  None
  * @retval None
  */
//<Added>
void TimingDelay_Decrement();
//</Added>

void SysTick_Handler(void)
{
  TimingDelay_Decrement();
}
```

The example I ended up with worked by putting some code in TimingDelay_Decrement(), which is defined in main.c

**Reply by earlephilhower** [reply]
>     1. To use the new cmsis compatbile macro in rename_exceptions.h, I need add cmsis_core to my project's CMake file to link this interface library. Can we add this in arduino-pico, i asusme there is no side effect.

I just tried it:
````
--- a/tools/libpico/CMakeLists.txt
+++ b/tools/libpico/CMakeLists.txt
@@ -67,6 +67,7 @@ target_link_libraries(pico
        pico_runtime
        pico_standard_link
        pico_stdlib
+       cmsis_core
        pico_unique_id
        tinyusb
         tinyusb_device_unmarked
````

The lib builds and some simple tests seem to work fine even afterwards.  I also see the symbols you want in the map file:
````
                0x00000000100001c8                PendSV_Handler
                0x00000000100001ca                SysTick_Handler
                0x0000000020000000                ram_vector_table
````

I suggest you use the `git` method to get a clone of the repo (see the README.md file) and try it yourself w/your code.

If that gets things going, let me know and I can update the main repo or you can submit a PR with the change.

**Reply by xiongyu0523** [reply]
Thanks I will try and follow-up.

**Reply by earlephilhower** [reply]
Oh yeah, just to be obvious, after adding the `cmsis_core` to the Cmakefile.txt, you need to run `./make-libpico.sh` to compile and install the libpico.a in the right spot...

**Reply by xiongyu0523** [reply]
The board package has included all required toolchain to build pico-sdk, right? I got below errors

```
./make-libpico.sh: line 6: cd: ../../system/arm-none-eabi/bin: No such file or directory
-- Building for: Visual Studio 16 2019
Using PICO_SDK_PATH from environment ('C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-sdk')
PICO_SDK_PATH is C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-sdk
Defaulting PICO_PLATFORM to rp2040 since not specified.
Defaulting PICO platform compiler to pico_arm_gcc since not specified.
-- Defaulting build type to 'Release' since not specified.
PICO compiler is pico_arm_gcc
Defaulting PICO_EXTRAS_PATH as sibling of PICO_SDK_PATH: C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-sdk/../pico-extras
-- The C compiler identification is MSVC 19.29.30133.0
-- The CXX compiler identification is MSVC 19.29.30133.0
-- The ASM compiler identification is GNU
-- Found assembler: C:/Program Files (x86)/GNU Arm Embedded Toolchain/10 2020-q4-major/bin/arm-none-eabi-gcc.exe
Defaulting PICO target board to pico since not specified.
Using board configuration from C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-sdk/src/boards/include/boards/pico.h
-- Found Python3: C:/Program Files/WindowsApps/PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0/python3.9.exe (found version "3.9.7") found components: Interpreter
TinyUSB available at C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-sdk/lib/tinyusb/src/portable/raspberrypi/rp2040; adding USB support.
lwIP available at C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/pico-extras/lib/lwip/src/core/tcp.c; TCP/IP support is available.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/Neo/AppData/Local/Arduino15/packages/rp2040/hardware/rp2040/1.9.4/tools/libpico/build
make: *** No targets specified and no makefile found.  Stop.

```

**Reply by earlephilhower** [reply]
You really need to use a GIT copy.  The Arduino IDE installs the binaries in a completely different folder.  Do the git clone and run `get.py` as described in `readme.md` then you can use the `makelibpico.sh` script which assumes a GIT copy to get the paths correct.

**Reply by xiongyu0523** [reply]
follow below steps i can get everything in `C:\Users\yuxiong\Arduino\hardware\pico\rp2040` folder. But how I can install in arduino-ide? I tried to copy rp2040 folder to my arduino package folder `C:\Users\yuxiong\Documents\ArduinoData\packages` where other ide installed  board package located but it does not works (can't find board package in IDE)

```
mkdir -p ~/Arduino/hardware/pico
git clone https://github.com/earlephilhower/arduino-pico.git ~/Arduino/hardware/pico/rp2040
cd ~/Arduino/hardware/pico/rp2040
git submodule update --init
cd pico-sdk
git submodule update --init
cd ../pico-extras
git submodule update --init
cd ../tools
python3 ./get.py
```

**Reply by xiongyu0523** [reply]
Hi. I'm still looking for a solution for this: )

**Reply by earlephilhower** [reply]
For Windows, the path is actually  `C:\Users\<user>\Documents\Arduino\hardware/pico/rp2040`.  (Assuming you used the .EXE installer and not the MSFT Windows Store version for Arduino which has a faked filesystem and really isn't recommended.)

When placed there, restarting the Arduino IDE should make it appear.  You can also add in the cmsys_core to the cmake file and rebuild before running the IDE to get the new symbols right off the bat.

**Comment by xiongyu0523** [comment]
Sorry, back to this topic after one year, I still want to add ThreadX to Arduino and RP2040. 

I just realized the pico-SDK was built into a static library in Arduino, what I need is to takeover systick and pendsv handler and override them with OS port defined in .S file, whether they are cmiss-compatible symbols (need link cmsis_core interface library and this is preferred to keep minimal code change) or private symbol (use isr_systick/isr_pendsv) in ctr0.S

Any suggestions?

**Comment by xiongyu0523** [comment]
More inputs:

- [ X ] ram_vector_table, I can access vector table from this symobl, no problem. 

- [ ? ] Need use ThreadX cortex-m0 port's PendSV_Handler to override the pico-sdk's weak dummy function. Prefer to keep the port unchanged so exposed symbol name should be PendSV_Handler and can be override. Not sure we can do this if you already build pico-sdk into .a. 
https://github.com/azure-rtos/threadx/blob/b871c33620f9b58a83e92a33f4ca00025e850c7f/ports/cortex_m0/gnu/src/tx_thread_schedule.S#L125

- [ ? ] Usually ThreadX scheduler and port implement tick timer in Systick of Cortex-M. In other Arudino platform like samd or stm32, I can find a hook function provide by the platform, so there is no need to override the systick ISR which is already used by the platform. In arduino-pico, it seems (when FreeRTOS is not used), the systick was used to provide epoch count but no hook entry is provided. Can we implement similiar things?
https://github.com/stm32duino/Arduino_Core_STM32/blob/ed4ef2f6ebbafd79b8aa13733397d2aa2ffbefd4/libraries/SrcWrapper/src/stm32/clock.c#L69
https://github.com/arduino/ArduinoCore-samd/blob/89833081ef18aea430b8943ab653dbbd510d43ce/cores/arduino/cortex_handlers.c#L172

last but not least, i think threadX should not work with FreeRTOS enabled, how can I turn off/on FreeRTOS in configuration?

**Reply by earlephilhower** [reply]
> ...PendSV handler...

FreeRTOS already overrides this in SW, (see their rp2040 port.c)

> ...Systick exception override...

I'm fine w/a PR to support this and it's pretty trivial to do, yes?  In fact, we overrive/ignore depending on the FreeRTOS linking state so the same logic could be done here.  See the `__isFreeRTOS` global variable's definition/setup.

> how can I turn off/on FreeRTOS in configuration

It's already off by default.  The user needs to manually add `#include <FreeRTOS.h>` to their code, OTW it's not used at all in the core.

**Reply by xiongyu0523** [reply]
Thanks.

- FreeRTOS port override Systick and PendSV by using exception_set_exclusive_handler() API offered by pico-sdk, this is a platform dependent call so I need make the bsp specific code to do this.  Currently for other platform there is no need to do this, just override the PendSV_Handler symbol with a strong symbol defined in standard cortex-m0 port. If I can do the same thing with arduino-pico then the overall porting layer can be keep the same. 

- Yes I think such a hook for application is very useful for people who want to reuse systick. I don't quite understand the code with _epoch, can I re-configure the systick rate (maybe 100Hz) and does not affect the original purpose of using systick in arudino core?

**Reply by earlephilhower** [reply]
For the `SystickPendSV` handlers, please really consider using the SDK calls and `#ifdef ARDUINO_ARCH_RP2040` to bracket it.  We don't want to hack the SDK to remove expected functionality.

Systick is used to give a CPU cycle count, which is useful for profiling.  There is a fallback PIO based counter, used by FreeRTOS, which you can use.  Look at the `RP2040Support.h` file where it shows how to register or fallback.  Weak linkage would make it simple enough.

**Comment by xiongyu0523** [comment]
Thanks again.

I was able to call the SDK API to register PendSV in the low level initialization function in ASM. I checked the right function address is placed into ram_vector_table[14], which is used by NVIC as vector. So I think PendSV will be fine.

```asm
  PUSH    {r0, lr}
  LDR     r0, =-2
  LDR     r1, =PendSV_Handler
  BL      exception_set_exclusive_handler
  POP     {r0, r1}
  MOV     lr, r1
```

I'm still trying to figure out how to do with Systick. I think my project will not compatible with FreeRTOS, and as you said, by default it is off. So PIO is not used. You have below code to register a callback `_SystickHandler()` to increment `_epoch`. 

```c
if (!__isFreeRTOS) {
    // Enable SYSTICK exception
    exception_set_exclusive_handler(SYSTICK_EXCEPTION, _SystickHandler);
    systick_hw->csr = 0x7;
    systick_hw->rvr = 0x00FFFFFF;
} else {
```

```c
private:
  static void _SystickHandler() {
      rp2040._epoch += 1LL << 24;
  }
```

Can I override this systick and will system works just fine? Because when I register my callback to replace yours, my rp2040 looks dead and USB is not funcitoning, I try to add some blink code in my task and it seems ThreadX multitask environement is not running. 

```
    PUSH    {r0, lr}
    LDR     r0, =-1
    LDR     r1, =SysTick_Handler
    BL      exception_set_exclusive_handler
    POP     {r0, r1}
    MOV     lr, r1
```
Or I should not override your _SystickHandler? It's quite hard to debug in arduino environment without a debugger.

**Comment by xiongyu0523** [comment]
Just FYI. my port of ThreadX is verified and works on a single cortex-m0 configuration on rp2040.

**Reply by earlephilhower** [reply]
Cool! 

I think the systick stuff is simpler than you think.    Add a bool `__isThreadX` and define it off of the existence of a weak function you have in your code.

Then at https://github.com/earlephilhower/arduino-pico/blob/98c4b921b81480809b60b3df8dbeb9fc904f31f2/cores/rp2040/RP2040Support.h#L224 and https://github.com/earlephilhower/arduino-pico/blob/98c4b921b81480809b60b3df8dbeb9fc904f31f2/cores/rp2040/RP2040Support.h#L252 and https://github.com/earlephilhower/arduino-pico/blob/98c4b921b81480809b60b3df8dbeb9fc904f31f2/cores/rp2040/RP2040Support.h#L266 you should check for `__isFreeRTOS || __isThreadX` and it will still give cycle counts using the PIO hw.

**Comment by xiongyu0523** [comment]
Thanks for your guidance.  I was able to get ThreadX environment running on top of Arduino-pico platform under specific configuration/modification. I also observed some strange behaviors that need further debug/discuss.

1) I think it does not make much sense to define __isThreadX in the core because eventually the core should not aware of using ThreadX, the PIO seems to be a better solution than systick for getting cpu cycles, so the platform can free systick for FreeRTOS, ThreadX and whatever use cases customer may want to use systick, so may I request to use PIO as always?

> Actually I tried to add _isThreadX and force it to true in the class for above three places, it works well but when I follow the exact what you did for _isFreerRTOS, define in main.cpp and delcare in Arduino.h and _freertos.h, the system seems not working.. I don't know why.

```c
class RP2040 {
public:
    RP2040()  { /* noop */ }
    ~RP2040() { /* noop */ }

    void begin() {
        _epoch = 0;
        int off = 0;
        _ccountPgm = new PIOProgram(&ccount_program);
        _ccountPgm->prepare(&_pio, &_sm, &off);
        ccount_program_init(_pio, _sm, off);
        pio_sm_set_enabled(_pio, _sm, true);
    }

   ......

    inline uint32_t getCycleCount() {
            return ccount_read(_pio, _sm);
    }

    inline uint64_t getCycleCount64() {
            return ccount_read(_pio, _sm);
    }

```

2) Another strange behavior the USB print is not stable when ThreadX is used. I can only get USB works by adding pinMode(LED_BUILTIN, OUTPUT) in setup(), otherwise, USB is not enum and board seems dead. 

```c
void setup()
{
    pinMode(LED_BUILTIN, OUTPUT);   //     If I move this line, USB is not recognized by windows, and board seems dead. 
    Serial.begin(115200);
    Serial.print("\r\nArduino Environment is Up\r\n");

    /* Enter the ThreadX kernel.  */
    tx_kernel_enter();
}
```

The library I made for testing is here, appreciate if you can have a try..
https://github.com/xiongyu0523/AzureRTOS-ThreadX-For-Arduino/tree/rp2040

**Reply by earlephilhower** [reply]
I haven't looked at your code, but think you have several issues with the port as-is from your description.

The core really does need to know that it's running multi-threaded and not just multi-core, and a lot of internal infrastructure needs to be adjusted just as was done for the FreeRTOS port:

* Per-thread reent data needs to be swapped in/out
* All mutexes/etc. need to be updated to use the proper ThreadX versions...now owner id == core Id, but in threaded models this is not true
* USB update should probably be a proper task, not a 1ms alarm, just like in FreeRTOS
* For flash updates anywhere (eeprom, filesystem, etc.), the core locking and the PendSV swap disable needs to be stopped (again, see the FreeRTOS port).

W/o those changes your port may work sometimes, but it won't be very stable.

**Reply by xiongyu0523** [reply]
Thank you I fully understand and agree with your comments.. These are further steps need to be done to deeply integrate with (any) arduino platform code to support all system-level features running a RTOS alongside. Current my plan with the ThreadX arduino library is generic and only provide basic CPU time slicing capabilities to use multithreading. So the first thing need to do is at least *IT IS WORKING* with minimum features, then I can see how much can invest on more support.

Now after move USB Serial initlization code into the initialization task (rather than in main), everythings works fine and stable now.

What I was asking about using PIO and avoid to occupy Systick in the core, is that good for you to make such a change? Any side effect you would be concern about? After this I can make first release and other may leverage the port to have engagement in the maker community just like this:
https://www.hackster.io/485734/azure-rtos-threadx-for-arduino-101-threads-963a8d

**Reply by earlephilhower** [reply]
Understood, but I would warn you that leaving those changed undone will result in very hard to debug problems later on (believe me, I know since I did the same thing for FreeRTOS and spent quite a while tracking things down!)

For SysTick, I really do not want to dedicate a PIO unit to a cycle counter.  There is a limited number of PIO units and program memory, and there are some people using the SW USB that need every single word of PIO memory already.  Systick is there and unused in the main SDK, and there's a simple way to fallback to PIO mode if systick is used elsewhere (like in FreeRTOS).

**Reply by xiongyu0523** [reply]
I'm looking into other options. 
1) Provide a weak hook() in your systick timer ISR is not ok with current code because you expect the timer to be expired at maximum duration for _epoch alogrithm, unless rewrite that part to be compatbile with any frequency (or 10ms) duration. I assume this will more complex for you to do or please correct me. 

2) Add __isAzureRTOS just same as __isFreeRTOS. I have some trobules. 

Here are modifications in `RP2040Support.h`

```c
    void begin() {
        _epoch = 0;
        if (!(__isFreeRTOS || __isAzureRTOS)) {
            // Enable SYSTICK exception
            exception_set_exclusive_handler(SYSTICK_EXCEPTION, _SystickHandler);
            systick_hw->csr = 0x7;
            systick_hw->rvr = 0x00FFFFFF;
        } else {
```

```c
    inline uint32_t getCycleCount() {
        if (!(__isFreeRTOS || __isAzureRTOS)) {
            uint32_t epoch;
            uint32_t ctr;
            do {
                epoch = (uint32_t)_epoch;
                ctr = systick_hw->cvr;
            } while (epoch != (uint32_t)_epoch);
            return epoch + (1 << 24) - ctr; /* CTR counts down from 1<<24-1 */
        } else {
            return ccount_read(_pio, _sm);
        }
    }

    inline uint64_t getCycleCount64() {
        if (!(__isFreeRTOS || __isAzureRTOS)) {
            uint64_t epoch;
            uint64_t ctr;
            do {
                epoch = _epoch;
                ctr = systick_hw->cvr;
            } while (epoch != _epoch);
            return epoch + (1LL << 24) - ctr;
        } else {
            return ccount_read(_pio, _sm);
        }
    }
```

And `main.cpp`
If I define *__isAzureRTOS* variable and initialize it to true in main, the USB won't work after downloading.
```c
   bool __isAzureRTOS = true;   // just after bool __isFreeRTOS;

    // Let rest of core know if we're using FreeRTOS
    __isFreeRTOS = initFreeRTOS ? true : false;
    __isAzureRTOS = true;
```
But if I initialize the variable in definition, everything will be ok. 
```c
bool __isAzureRTOS = true;
```

I couldn't figure out what's going on there. Any ideas?  

Btw, I think think systick is more important resources than PIO because using the chip has one only single instance and it is common.

**Reply by earlephilhower** [reply]
Sorry, for the USB bit you're breaking something and will need to look at GDB or a trace to see what's failing.

I'd do the `tusb_init()` before the threadx init, and then later I'd make a threadx process that does the tusb process call followed by a 1ms sleep in an infinite loop.  I would not use the alarm-based interrupt driven update.  Again, I strongly recommend copying the FreeRTOS method since you're doing essentially the same thing.

The PIO counter override trivially just starts the PIO program which runs w/o **any** core usage whatsoever until you manually read it.

Basically, I think it all comes down to the USB being busted by the threadx code, but w/o tracing it's hard to guess what's wrong.  The FreeRTOS methods in the core for USB setup/run are working well so please implement the same in threadx and I have a feeling you'll be in good shape.

**Reply by earlephilhower** [reply]
Actually, now that I think about it, you were using a call in the main app `setup()` to start threadx yes?

That is definitely is going to kill everything.  The USB stack is interrupt driven by default and has timing restrictions in some bits.  Your OS code is using PendSV to swap tasks, which **overrides** the interrupt state IIRC, meaning the USB core could be in the process of being processed and then suddenly swapped out.  Not only would this be a timing problem, the rest of the core can't use a USB stack that's in an incoherent state so you could end up crashing on a `Serial` write/read even if the main swapped-out IRQ doesn't crash the system.

**Comment by xiongyu0523** [comment]
Oh yes. Actually the when I start ThreadX environmenet in `setup()`, it should never return.  I see (when FreeRTOS is not used) the tinyUSB stack is fully interrupt-driven and also handled in the USB ISR. I'm not very sure what are doing in __loop() function at the background, it seems serialEvent() is not defined and has no effect currently? If it does some meaningful then my integration will have big problem because it will never be called since it is after `setup()`. 

```c
void arduino::serialEventRun(void) {
    if (serialEvent && Serial.available()) {
        serialEvent();
    }
}
```

PendSV should be the least priority configured by the initalization code, so I'm not sure how it will affect USB or other interrupt. In RTOS based system, urgent/frequent/time-sensing interrupt could be run independ of RTOS as long as user do not call RTOS specific API or use RTOS data in that ISR. That's my understanding..

But anyway, I see a deep integration of FreeRTOS in your core, I probably need to think about whether it is worth and how can do this in a similiar way.. But few things want to double check with you before start. 

1) ThreadX port does not support SMP/Dual core, as long as I do not call setup1() and/or loop1() in the .ino, the core infrasture code will not enable SMP/dual core and I can stick with single cortex-m0 config, right?
2) My library is for multi platform and not directly integrated into arduino-core, is it different from your built-in libraries anyway?

Appreciate!

**Reply by earlephilhower** [reply]
> ...serialEvent() is not defined...

`serialEvent` is an old Arduino standard call that uses `weak` functions.  It's not common at all now, but it is part of the docs/etc. so that it why it's in the internal __loop.  You've broken compatibility with that change, but it sounds like your library doesn't even call `loop()` so all Arduino stuff is busted at that point, anyway.

> I see a deep integration of FreeRTOS in your core, I probably need to think about whether it is worth and how can do this in a similiar way.

If you are taking over control of the processor and interrupt handling and adding threads, you have no choice but to integrate it or it will be not useful at all.  `mutex`es no longer work with your library which breaks pretty much everything.  Interrupts can be swapped out, etc.  This processor is a little more complicated than an 8-bit AVR. :)

> as long as I do not call setup1() and/or loop1() in the .ino, the core infrasture code will not enable SMP/dual core and I can stick with single cortex-m0 config, right?

Yes

> My library is for multi platform and not directly integrated into arduino-core, is it different from your built-in libraries anyway?

The FreeRTOS SMP lib included here is the upstream project plus some minor fixes to let it integrate properly and make the rest of the Arduino stuff work properly.  I can't imagine how a different threading library doing essentially the same thing could work without the same kinds of changes.

At a very basic level, any multithreading **breaks a basic assumption of the Pico-SDK mutex**.  The owner of the mutex is the core ID, but now you can have the same core ID but different processes so you need to not use the SDK mutexes and implement your own, for example.  There are other real-world issues, too.

**Reply by earlephilhower** [reply]
Here's an idea:  Since ThreadX isn't SMP safe/supported, why not avoid the whole issue and run ThreadX on `core1` and leave `core0` for all the standard Arduino stuff?

That way you basically have your own system with no interrupts, no exceptions, etc. from other code that need to run.  Everything running on core1 would be your own code...
